feat: Add support for push notifications via `ParsePush`, `ParseNotification` (#914)
parse-community/Parse-SDK-Flutter: packages/flutter/PUSH.md

# Push Notifications

Push notifications are a great way to keep your users engaged and informed about your app. You can reach your user base quickly and effectively. This guide will help you through the setup process and the general usage of Parse Platform to send push notifications.

To configure push notifications in Parse Server, check out the [push notification guide](https://docs.parseplatform.org/parse-server/guide/#push-notifications).

## Installation

1. Install [Firebase Core](https://firebase.flutter.dev/docs/overview) and [Cloud Messaging](https://firebase.flutter.dev/docs/messaging/overview). For more details review the [Firebase Core Manual](https://firebase.flutter.dev/docs/manual-installation/).

2. Add the following code after `Parse().initialize(...);`:

```dart
ParsePush.instance.initialize(FirebaseMessaging.instance);

FirebaseMessaging.onMessage.listen((message) => ParsePush.instance.onMessage(message));

> ```

Reviewer: Why has the indentation been removed?

3. For you app to process push notification while in the background, add the following code:
```dart
FirebaseMessaging.onBackgroundMessage(onBackgroundMessage);
```

```dart
Future<void> onBackgroundMessage(RemoteMessage message) async => ParsePush.instance.onMessage(message);

> ```

Reviewer: Why has the indentation been removed?

# Push Notifications

> Push notifications are a great way to keep your users engaged and informed about your app. You can reach your entire user base quickly and effectively. This guide will help you through the setup process and the general usage of Parse to send push notifications.

Reviewer: ```suggestion
Push notifications are a great way to keep your users engaged and informed about your app. You can reach your user base quickly and effectively. This guide will help you through the setup process and the general usage of Parse Platform to send push notifications.
```

> To activate and implement Push Notifications in Parse Server, check [this page](https://docs.parseplatform.org/parse-server/guide/#push-notifications)

Reviewer: ```suggestion
To configure push notifications in Parse Server, check out the [push notification guide](https://docs.parseplatform.org/parse-server/guide/#push-notifications).
```

> ## Installation
> 1 : First need install [Firebase Core](https://firebase.flutter.dev/docs/overview) and [Cloud Messaging](https://firebase.flutter.dev/docs/messaging/overview)
> 
> Tip : Recommend reviewing the [Firebase Core Manual](https://firebase.flutter.dev/docs/manual-installation/)
> 
> 2 : Set the following codes after ```Parse().initialize```
> ```dart
> await Parse().initialize(...);
> 
> ParsePush.instance.initialize(FirebaseMessaging.instance);
> FirebaseMessaging.onMessage.listen((message) => ParsePush.instance.onMessage(message));
> ```
> 
> 3 : To work with push notifications after closing the application, follow the steps below.
> 
> Put the following code after the above codes
> ```dart
> FirebaseMessaging.onBackgroundMessage(onBackgroundMessage);
> ```
> 
> Put the following function in your codes
> ```dart
> Future<void> onBackgroundMessage(RemoteMessage message) async => ParsePush.instance.onMessage(message);
> ```

Reviewer: ```suggestion
## Installation

1. Install [Firebase Core](https://firebase.flutter.dev/docs/overview) and [Cloud Messaging](https://firebase.flutter.dev/docs/messaging/overview). For more details review the [Firebase Core Manual](https://firebase.flutter.dev/docs/manual-installation/).

2. Add the following code after `Parse().initialize(...);`:

  ```dart
  ParsePush.instance.initialize(FirebaseMessaging.instance);
  FirebaseMessaging.onMessage.listen((message) => ParsePush.instance.onMessage(message));
  ```

3. For you app to process push notification while in the background, add the following code:

  ```dart
  FirebaseMessaging.onBackgroundMessage(onBackgroundMessage);
  ```

  ```dart
  Future<void> onBackgroundMessage(RemoteMessage message) async => ParsePush.instance.onMessage(message);
  ```

Author: Do you think the Installation section was too short? Of course it is understandable!

> ## Implemented example
> Your code should look like the following code

Reviewer: ```suggestion
## Implementation Example

The following is a code example for a simple implementation of push notifications:
```

```dart

> Future<void> main() async {
>   WidgetsFlutterBinding.ensureInitialized();
>   
>   // initialize Firebase Core
>   await Firebase.initializeApp(
>     options: DefaultFirebaseOptions.currentPlatform,
>   );
> 
>   // initialize Parse
>   await Parse().initialize("applicationId", "serverUrl",
>       clientKey: "clientKey", debug: true);
> 
>   // initialize Parse Push
>   ParsePush.instance.initialize(FirebaseMessaging.instance);
>   FirebaseMessaging.onMessage
>       .listen((message) => ParsePush.instance.onMessage(message));
>   
>   // for run ParsePush in the background
>   FirebaseMessaging.onBackgroundMessage(onBackgroundMessage);
> 
>   runApp(const MyApp());
> }

Reviewer: ```suggestion
Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  
  // Initialize Firebase Core
  await Firebase.initializeApp(
    options: DefaultFirebaseOptions.currentPlatform,
  );

  // Initialize Parse
  await Parse().initialize("applicationId", "serverUrl",
      clientKey: "clientKey", debug: true);

  // Initialize Parse push notifications
  ParsePush.instance.initialize(FirebaseMessaging.instance);
  FirebaseMessaging.onMessage
      .listen((message) => ParsePush.instance.onMessage(message));
  
  // Process push notifications while app is in the background
  FirebaseMessaging.onBackgroundMessage(onBackgroundMessage);

  runApp(const MyApp());
}
```
